linked_list/linked_list.py

```diff
@@ -11,69 +11,116 @@ class LinkedList:
     def __init__(self):
       self.head = None # keep the head private. Not accessible outside this class
 
-    # returns the value in the first node
-    # returns None if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first(self):
-        pass
-
+        if self.head == None:
+            return None
+        value = self.head.value
+        return value
 
-    # method to add a new node with the specific data value in the linked list
-    # insert the new node at the beginning of the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_first(self, value):
-        pass
-
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+        new_node = Node(value)
+        new_node.next = self.head
+        self.head = new_node
+        
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def search(self, value):
-        pass
-
-    # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+        if self.head == None:
+            return False
+        current = self.head
+        while current is not None:
+            if current.value == value:
+                return True
+            current = current.next
+        return False
+            
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def length(self):

```

Comment:
👍 Note you could also maintain a `length` attribute and update it when you add and remove nodes.  Then you could do this method in O(1) time complexity.